juju/juju
# WIP: 2.2 prune presence 1454661

## Description of change

This adds a "BeingPruner" to state/presence. This will clean up presence.beings and presence.pings so that they only contain recent sequences. Any Key that gets superseded by another sequence no longer pays attention to the old sequence, so it is safe to prune that sequence from the database. And pings older than 2 slots are not used either, so it is safe to remove slots.

This also triggers running the pruner on a pseudo-random cycle. We Sync all Watchers every 30s, and this triggers a prune (on average) every 20 Syncs (every 10 minutes). We use psuedo-random so that in HA, etc, we don't end up in sync across all the watchers. I would consider either
a) Merging this into the txnPruner logic, so it strictly happens every 2hrs.
b) Turning down the frequency.

The other bit about psuedo-random is that it biases so that we are unlikely to run 2x in a row, and will guarantee that we will eventually trigger (each time it doesn't trigger increases the chance it will trigger next time).

I'm currently doing some live testing, and I added a test that does abusive testing. There isn't a lot of focused unit tests, which is why this is WIP.

One thing that is surprising me is that at steady state, we're still seeing "unknown" sequences for pingers. I haven't dug into what keys those represent, but with 180 active unit agents, 
```
machine-0: 18:43:02 DEBUG juju.state.presence looked up 3 unknown sequences for "c3525263-1fe3-4556-8b5c-2bbc3beee6f7" (3 unowned) in 805.143µs (3.7seq/ms) from "presence.beings"
```

That means that in the active 2 ping slots, there are 3 bits set that don't match a known-active agent/key. Now, we generate a new sequence whenever something reconnects, but nothing should be actively doing very much. (Unless we are generating presence sequences for client connections, as I have a 'watch juju status' running.)

This also refactors the Presence logic a lot, so we no longer need to pre-load the 'allBeings' map. Instead we batch-load all of the items that we aren't already actively tracking. At high number of active pingers this improves load fairly significantly. (The torture test creates 3500 active pingers and cycles them every couple of seconds, and loading that is improved significantly.)

## QA steps

Spin up a reasonable sized deployment, with quite a few agents. Run it for a while, make sure none of the agents appears to go offline incorrectly, and monitor the size of presence collections. They should stay at a reasonable size.

You can also currently use:
```
  juju debug-log -m controller --include-machine machine-0 --include-module juju.state.presence
```
To see what is going on. The "lookup" logging statement should probably drop to Tracef if it isn't going to stay silent during normal behavior.

Another trick to force load on the system is to restart various agents, so that they're forced to get new sequences. The easiest way to do that is to bounce controller agents, so everything connected to them has to reconnect.
```
  juju ssh $X sudo /usr/sbin/service jujud-machine-$X restart
```


## Documentation changes

Shouldn't be user-visible, other than improvements around responsiveness, etc for 'juju status'.

## Bug reference

[lp:1454661](https://bugs.launchpad.net/juju/+bug/1454661)
